File: leetcode3/최민지/312_Burst_Balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        # 1. 양 끝에 가상 풍선 1 추가 및 길이 정의
        A = [1] + nums + [1]
        n = len(A)

        # 2차원 dp 테이블 초기화 (dp[i][j] 사이의 풍선들을 모두 터뜨렸을 떄의 최대 점수 )
        dp = [[0] * n for _ in range(n)]

        # 구간의 길이(length)를 2부터 n-1까지 늘려가며 계산
        for length in range(2, n):
            for left in range(0, n-length):
                right = left + length

                # left와 right의 사이에서 '가장 마지막으로 터뜨릴 풍선 k'를 탐색
                for k in range(left + 1, right):
                    coins = dp[left][k] + dp[k][right] + (A[left] * A[k] * A[right])
                
                    if coins > dp[left][right]:
                        dp[left][right] = coins

        return dp[0][n-1]

File: leetcode3/황은지/312. Burst Balloons.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxCoins = function (nums) {
  // 모든 순열 조합으로 하면 너무 숫자가 커짐(300!)..
  // 단순히 생각햇을땐, 가장 작은수부터 빨리 터뜨려야 이후 연쇄 곱셉에 큰수가 같이 곱해짐
  let coins = 0;
  nums.sort((a, b) => a - b);
  const arrNum = [...nums];
  let p = 0;

  while (arrNum.length > 0) {
    const minNum = nums[p++];
    for (let i = 0; i < arrNum.length; i++) {
      if (arrNum[i] === minNum) {
        const leftNum = i === 0 ? 1 : arrNum[i - 1];
        const rightNum = i === arrNum.length - 1 ? 1 : arrNum[i + 1];
        arrNum.splice(i, 1);
        break;
      }
    }
  }
  return coins;
};
